File: client/src/lib/crdt/gcounter.test.ts
import GCounter from "./gcounter";
import { describe, expect, it } from "vitest";

describe("GCounter", () => {
    it("should be able to increment", () => {
        const counter = new GCounter();

        expect(counter.value).toBe(0);
        expect(counter.inc()).toBe(1);
    });

    it("should be able to increment multiple times", () => {
        const counter = new GCounter();

        expect(counter.value).toBe(0);
        expect(counter.inc()).toBe(1);
        expect(counter.inc()).toBe(2);
        expect(counter.inc()).toBe(3);
    });

    it("should be able to increment by any positive value", () => {
        const counter = new GCounter();

        expect(counter.value).toBe(0);
        expect(counter.inc(5)).toBe(5);
        expect(counter.inc(3)).toBe(8);
        expect(counter.inc(2)).toBe(10);
    });

    it("should not be able to increment by a negative value", () => {
        const counter = new GCounter();

        expect(counter.value).toBe(0);
        expect(() => counter.inc(-1)).toThrow();
    });

    it("should be able to start with a value", () => {
        const counter = new GCounter(5);

        expect(counter.value).toBe(5);
    });

    it("should be able to merge with itself", () => {
        const counter = new GCounter();

        counter.inc();
        counter.inc();

        expect(counter.merge(counter)).toBe(2);
    });

    it("should be able to merge with another counter", () => {
        const counter1 = new GCounter();
        const counter2 = new GCounter();

        counter1.inc();
        counter1.inc();
        counter2.inc();

        expect(counter1.merge(counter2)).toBe(2);
    });
});

File: server/src/test/kotlin/pt/up/fe/sdle/crdt/GCounterTest.kt
package pt.up.fe.sdle.crdt

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFails

class GCounterTest {
    @Test
    fun `GCounter - increment`() {
        val counter = GCounter()

        assertEquals(0, counter.value)
        assertEquals(1, counter.inc())
    }

    @Test
    fun `GCounter - increment multiple times`() {
        val counter = GCounter()

        assertEquals(0, counter.value)
        assertEquals(1, counter.inc())
        assertEquals(2, counter.inc())
        assertEquals(3, counter.inc())
    }

    @Test
    fun `GCounter - increment by any positive value`() {
        val counter = GCounter()

        assertEquals(0, counter.value)
        assertEquals(5, counter.inc(5))
        assertEquals(8, counter.inc(3))
        assertEquals(10, counter.inc(2))
    }

    @Test
    fun `GCounter - increment by a negative value`() {
        val counter = GCounter()

        assertEquals(0, counter.value)
        assertFails { counter.inc(-1) }
    }

    @Test
    fun `GCounter - start with a value`() {
        val counter = GCounter(5)

        assertEquals(5, counter.value)
    }

    @Test
    fun `GCounter - merge with itself`() {
        val counter = GCounter()

        counter.inc()
        counter.inc()

        assertEquals(2, counter.merge(counter))
    }

    @Test
    fun `GCounter - merge with another counter`() {
        val counter1 = GCounter()
        val counter2 = GCounter()

        counter1.inc()
        counter1.inc()
        counter2.inc()

        assertEquals(2, counter1.merge(counter2))
    }
}
